skeleton/tsh.1
.\" Process this file with
.\" groff -man -Tascii tsh.1
.\"
.TH TSH 1 "OCTOBER 2009" "NU EECS 343" "NU EECS 343 - Operating Systems - Fall 2009"
.SH NAME
tsh \- A tiny shell
.SH SYNOPSIS
.B tsh
.SH DESCRIPTION
.B tsh
is a very basic shell implementation.  It has three built-in commands, described below.  All other commands
entered are assumed to be references to some exectuable file, which tsh attempts to find by searching
the directories specified in the PATH environment variable.  If it can find a file in the path, it forks 
a child process and executes the file there.  Otherwise it reports an error message.  
.SH BUILT-IN COMMANDS
.IP exit
Quit tsh
.IP cd directory
Changes the current working directory to directory.
.IP VAR=value
Updates the value of the environment variable VAR to be value.
.SH DESIGN APPROACH
I took the path of least resistance and used the test cases to guide my development.  Most of the work went
into implementing functions specified in runtime.c, with a few changes to interpreter.c.  Child processes, once
they have been located, are forked in the Exec function.  The shell process then goes into a loop waiting for
SIGCHLD signals; the signal handler calls reap_children, which reaps all available child processes.  If the foreground
process is reaped, the global variable fgpid is set to -1, which causes the parent process to continue.

I changed the error message to satisfy test 7 as suggested in the Google group, and although test 11 fails I
assumed that was ok as indicated there.

I did a small amount of work toward extra credit.  Environment variables can be set and accessed (vie $VAR), and the
shell will open and execute commands form ~/.tshrc.  However, the output from ~/.tshrc is not suppressed and errors are not
handled nicely.  And I didn't do any work on the fancy prompt stuff.  I implemented the environment variables
through the built-in command infrastructure instead of the interpreter.

.SH AUTHOR
Burke Fetscher
baf111
burke.fetscher@eecs.northwestern.edu


skeleton/tsh.c
/***************************************************************************
 *  Title: tsh
 * -------------------------------------------------------------------------
 *    Purpose: A simple shell implementation 
 *    Author: Stefan Birrer
 *    Version: $Revision: 1.4 $
 *    Last Modification: $Date: 2009/10/12 20:50:12 $
 *    File: $RCSfile: tsh.c,v $
 *    Copyright: (C) 2002 by Stefan Birrer
 ***************************************************************************/
#define __MYSS_IMPL__

/************System include***********************************************/
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/types.h>

/************Private include**********************************************/
#include "tsh.h"
#include "io.h"
#include "interpreter.h"
#include "runtime.h"

/************Defines and Typedefs*****************************************/
/*  #defines and typedefs should have their names in all caps.
 *  Global variables begin with g. Global constants with k. Local
 *  variables should be in all lower case. When initializing
 *  structures and arrays, line everything up in neat columns.
 */

#define BUFSIZE 80

/************Global Variables*********************************************/

/************Function Prototypes******************************************/
/* handles SIGINT and SIGSTOP signals */
static void 
sig(int);
/* reaps all finished child processes */
static void 
reap_children();
/* processes the tshrc file */
static void 
process_tshrc();
/************External Declaration*****************************************/

/**************Implementation***********************************************/

/*
 * main
 *
 * arguments:
 *   int argc: the number of arguments provided on the command line
 *   char *argv[]: array of strings provided on the command line
 *
 * returns: int: 0 = OK, else error
 *
 * This sets up signal handling and implements the main loop of tsh.
 */
int
main(int argc, char *argv[])
{
  /* Initialize command buffer */
  char* cmdLine = malloc(sizeof(char*) * BUFSIZE);

  /* shell initialization */
  if (signal(SIGINT, sig) == SIG_ERR)
    PrintPError("SIGINT");
  if (signal(SIGTSTP, sig) == SIG_ERR)
    PrintPError("SIGTSTP");
  if (signal(SIGCHLD, sig) == SIG_ERR)
    PrintPError("SIGCHLD");

  fgpid = -1;

  process_tshrc();

  while (!forceExit) /* repeat forever */
    {
      /* print prompt */
      //printf("tsh>");
      
      /* read command line */
      getCommandLine(&cmdLine, BUFSIZE);


      if (strcmp(cmdLine, "exit") == 0) {
	//printf("quitting...\n");
        forceExit = TRUE;
	fflush(stdout);
	continue;
      }

      /* checks the status of background jobs */
      CheckJobs();
 
      /* interpret command and line
       * includes executing of commands */
      Interpret(cmdLine);

    }

  /* shell termination */
  free(cmdLine);
  
  return 0;
} /* main */

/*
 * sig
 *
 * arguments:
 *   int signo: the signal being sent
 *
 * returns: none
 *
 * This should handle signals sent to tsh.
 */
static void
sig(int signo)
{
  if (signo == SIGINT) 
    IntFgProc();
  if (signo == SIGTSTP)
    StopFgProc();
  if (signo == SIGCHLD)
    reap_children();
} /* sig */

/*
 * reap_children
 *
 * arguments: none
 *
 * returns: none
 *
 * Reap all stoped child processes.
 */
static void
reap_children()
{
  int status = 0;
  int pid = 0;
  do 
    {
      pid = waitpid(-1, &status, WNOHANG | WUNTRACED);
      // if fg process has stopped, update fgpid
      if (pid == fgpid)
	fgpid = -1;
      if (WIFSTOPPED(status)) {
	updatebgjob(pid, STOPPED);
      }
      else
	updatebgjob(pid, DONE);
    } while (pid > 0);
} /*reap_children */

/*
 * process_tshrc
 *
 * arguments: none
 *
 * returns: none
 *
 * Read and process the tshrc file.
 */
static void 
process_tshrc()
{
  char *homedir;
  char *fpath;
  struct stat st;
  FILE *tshrc_file;
  const char *fname = ".tshrc";

  // set up path to ~/.tshrc
  homedir = getenv("HOME");
  fpath = (char *)malloc(2 + sizeof(char) * (strlen(homedir) + strlen(fname)));
  strcpy(fpath, homedir);
  strcat(fpath, "/");
  strcat(fpath, fname);

  // if ~/.tshrc exists, read it and interpret non-comment lines
  if (stat(fpath, &st) == 0) {
    tshrc_file = fopen(fpath, "r");
    char line[128];
    while (fgets(line, sizeof(line), tshrc_file) != NULL) {
      if (line[0] != '#') {
	Interpret(line);
      }
    }
    fclose(tshrc_file);
  }
  free(fpath);
}
